src/syntax/syntax.rkt

```diff
@@ -565,108 +426,68 @@
-                        binary64
-                        #:spec (sin x)
-                        #:fpcore (! :precision binary64 (sin x))
-                        #:fl sin)
-
-  (define log1pmd-proc (impl-info 'log1pmd.f64 'fl))
-  (define log1pmd-vals '((0.0 . 0.0) (0.5 . 1.0986122886681098) (-0.5 . -1.0986122886681098)))
-  (for ([(pt out) (in-dict log1pmd-vals)])
-    (check-equal? (log1pmd-proc pt) out (format "log1pmd(~a) = ~a" pt out)))
-
-  (define sin-proc (impl-info 'sin.acc.f64 'fl))
-  (define sin-vals '((0.0 . 0.0) (1.0 . 0.8414709848078965) (-1.0 . -0.8414709848078965)))
-  (for ([(pt out) (in-dict sin-vals)])
-    (check-equal? (sin-proc pt) out (format "sin(~a) = ~a" pt out)))
-
-  (void))
+;; Extracts the `fpcore` field of an operator implementation
+;; as a property dictionary and expression.
+(define (impl->fpcore impl)
+  (match (impl-info impl 'fpcore)
+    [(list '! props ... body) (values (props->dict props) body)]
+    [body (values '() body)]))
+
+;; For a given FPCore operator, rounding context, and input representations,
+;; finds the best operator implementation. Panics if none can be found.
+(define/contract (get-fpcore-impl op prop-dict ireprs #:impls [all-impls (all-active-operator-impls)])
+  (->* (symbol? prop-dict/c (listof representation?)) (#:impls (listof symbol?)) symbol?)
+  ; gather all implementations that have the same spec, input representations,
+  ; and its FPCore translation has properties that are found in `prop-dict`
+  (define impls
+    (reap [sow]
+          (for ([impl (in-list all-impls)])
+            (when (equal? ireprs (impl-info impl 'itype))
+              (define-values (prop-dict* expr) (impl->fpcore impl))
+              (define pattern (cons op (map (lambda (_) (gensym)) ireprs)))
+              (when (and (andmap (lambda (prop) (member prop prop-dict)) prop-dict*)
+                         (pattern-match pattern expr))
+                (sow impl))))))
+  ; check that we have any matching impls
+  (when (null? impls)
+    (raise-herbie-missing-error
+     "No implementation for `~a` under rounding context `~a` with types `~a`"
+     op
+     prop-dict
+     (string-join (map (λ (r) (format "<~a>" (representation-name r))) ireprs) " ")))
+  ; ; we rank implementations and select the highest scoring one
+  (define scores
+    (for/list ([impl (in-list impls)])
+      (define-values (prop-dict* _) (impl->fpcore impl))
+      (define num-matching (count (lambda (prop) (member prop prop-dict*)) prop-dict))
+      (cons num-matching (- (length prop-dict) num-matching))))
+  ; select the best implementation
+  ; sort first by the number of matched properties,
+  ; then tie break on the number of extraneous properties
+  (match-define (list (cons _ best) _ ...)
+    (sort (map cons scores impls)
+          (lambda (x y)
+            (cond
+              [(> (car x) (car y)) #t]
+              [(< (car x) (car y)) #f]
+              [else (> (cdr x) (cdr y))]))
+          #:key car))
```

Comment:
Does this ever actually happen, that we match multiple? I guess for something like `expm1` we want to match that and not `exp` and subtraction?